File: ios/Cindy/Signal/EMAFilter.swift
import Foundation

/// Exponential moving average. `alpha` = 1 passes the input through unchanged.
struct EMAFilter: Sendable {
    let alpha: Float
    private(set) var value: Float?

    init(alpha: Float) {
        self.alpha = alpha
    }

    /// Feeds one sample and returns the smoothed value.
    @discardableResult
    mutating func update(_ x: Float) -> Float {
        if let v = value {
            let next = v + alpha * (x - v)
            value = next
            return next
        }
        value = x
        return x
    }

    mutating func reset() {
        value = nil
    }
}

File: android/core/src/main/kotlin/me/raddatz/cindy/core/signal/EMAFilter.kt
package me.raddatz.cindy.core.signal

/** Exponential moving average. `alpha` = 1 passes the input through unchanged. */
class EMAFilter(val alpha: Float) {
    var value: Float? = null
        private set

    /** Feeds one sample and returns the smoothed value. */
    fun update(x: Float): Float {
        val v = value
        if (v != null) {
            val next = v + alpha * (x - v)
            value = next
            return next
        }
        value = x
        return x
    }

    fun reset() {
        value = null
    }
}
